src/environment.js

```diff
@@ -1382,17 +1382,26 @@ export class Reference {
     );
     this.base = base;
     this.referencedName = refName;
-    invariant(
-      !(refName instanceof AbstractValue) ||
-        !(
-          refName.mightNotBeString() &&
-          refName.mightNotBeNumber() &&
-          !refName.isSimpleObject() &&
-          // if the base is a simple abstract object but
-          // the refName is not simple, this is also okay
-          (base instanceof AbstractValue && !base.isSimpleObject())
-        )
-    );
+
+    if (
+      refName instanceof AbstractValue &&
+      (refName.mightNotBeString() &&
+        refName.mightNotBeNumber() &&
+        !refName.isSimpleObject() &&
+        // if the base is a simple abstract object but
+        // the refName is not simple, this is also okay
+        (base instanceof AbstractValue && !base.isSimpleObject()))
+    ) {
+      let error = new CompilerDiagnostic(
+        "cannot create an object property reference when the property is abstract",
+        base.$Realm.currentLocation,
```

Comment:
In general, it is better to issue diagnostics in the evaluator where the AST is close at hand. Realm.currentLocation could be wildly wrong. The purpose of the invariant is to require the caller to have done the necessary checks and to have issued the appropriate diagnostics.